linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala

```diff
@@ -709,7 +777,28 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
           args.forall(isTriviallySideEffectFree(_))
 
         case LoadModule(className) =>
-          dependenciesBuilder += className
+          isLoadModuleSideEffectFree(className)
```

Comment:
Under compliant module init semantics, the first `LoadModule` returns `null`. It throws if they're `Fatal` and (as an implementation detail) stack-overflows under `Unchecked`.

The way the current code is written interprets a `LoadModule(self)` before `StoreModule` as a cyclic dependency on oneself. It seems to me that it's the best interpretation, but we could also make it directly non-side-effect-free.

Reply:
> It seems to me that it's the best interpretation

Agreed.

> we could also make it directly non-side-effect-free.

Nevertheless, I think I'd prefer this approach. It makes it explicit that a self load must be handled and how it is handled. I've completely missed this possibility in the review of #4938.

Reply:
Another possibility would be to change the codegen to emit `This()` instead of `LoadModule(self)` when in the constructor and after the `StoreModule`. Then we wouldn't need to deal with `LoadModule(self)` any differently than other `LoadModule`s here. They would register a dependency on self, which the algorithm would then declare non elidable because cyclic. Fewer special cases.

Reply:
And then we'd disallow `LoadModule(self)` via `ClassDefChecker`? Seems much more complicated than adding an `if` here :-/

Reply:
No that would still be allowed. It would just be less optimized.